// graphql-code-generator-dart/dist/loaders/schema/schema-from-typedefs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var isGlob = require("is-glob");
var isValidPath = require("is-valid-path");
var glob = require("glob");
var graphql_tools_1 = require("graphql-tools");
var fs_1 = require("fs");
var graphql_import_1 = require("graphql-import");
var SchemaFromTypedefs = /** @class */ (function () {
    function SchemaFromTypedefs() {
    }
    SchemaFromTypedefs.prototype.canHandle = function (globPath) {
        return isGlob(globPath) || (isValidPath(globPath) && globPath.endsWith('.graphql'));
    };
    SchemaFromTypedefs.prototype.handle = function (globPath, cliOptions) {
        var globFiles = glob.sync(globPath, { cwd: process.cwd() });
        if (!globFiles || globFiles.length === 0) {
            throw new Error("Unable to find matching files for glob: " + globPath + "!");
        }
        var typeDefs = globFiles.length > 1 ? globFiles.map(function (filePath) { return fs_1.readFileSync(filePath, 'utf-8'); }) : graphql_import_1.importSchema(globFiles[0]);
        return graphql_tools_1.makeExecutableSchema({
            typeDefs: typeDefs,
            allowUndefinedInResolve: true,
            resolvers: {}
        });
    };
    return SchemaFromTypedefs;
}());
exports.SchemaFromTypedefs = SchemaFromTypedefs;
